arrow/arrow.py

```diff
@@ -1445,84 +1558,140 @@ def for_json(self):
         return self.__add__(other)
 
-    def __sub__(self, other):
+    @overload
+    def __sub__(self, other: Union[timedelta, relativedelta]) -> "Arrow":
+        pass  # pragma: no cover
+
+    @overload
+    def __sub__(self, other: Union[dt_datetime, "Arrow"]) -> timedelta:
+        pass  # pragma: no cover
+
+    def __sub__(self, other: Any) -> Union[timedelta, "Arrow"]:
 
         if isinstance(other, (timedelta, relativedelta)):
             return self.fromdatetime(self._datetime - other, self._datetime.tzinfo)
 
-        elif isinstance(other, datetime):
+        elif isinstance(other, dt_datetime):
             return self._datetime - other
 
         elif isinstance(other, Arrow):
             return self._datetime - other._datetime
 
         return NotImplemented
 
-    def __rsub__(self, other):
+    @overload
+    def __rsub__(self, other: dt_datetime) -> timedelta:
+        pass  # pragma: no cover
 
-        if isinstance(other, datetime):
+    @overload
+    def __rsub__(self, other: Any) -> timedelta:
+        pass  # pragma: no cover
+
+    def __rsub__(self, other: Any) -> timedelta:
+
+        if isinstance(other, dt_datetime):
             return other - self._datetime
 
         return NotImplemented
 
     # comparisons
 
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
 
-        if not isinstance(other, (Arrow, datetime)):
+        if not isinstance(other, (Arrow, dt_datetime)):
             return False
 
         return self._datetime == self._get_datetime(other)
 
-    def __ne__(self, other):
+    def __ne__(self, other: Any) -> bool:
 
-        if not isinstance(other, (Arrow, datetime)):
+        if not isinstance(other, (Arrow, dt_datetime)):
             return True
 
         return not self.__eq__(other)
 
-    def __gt__(self, other):
+    @overload
+    def __gt__(self, other: "Arrow") -> bool:
+        pass  # pragma: no cover
+
+    @overload
+    def __gt__(self, other: dt_datetime) -> bool:
+        pass  # pragma: no cover
 
-        if not isinstance(other, (Arrow, datetime)):
+    def __gt__(self, other: Any) -> bool:
+
+        if not isinstance(other, (Arrow, dt_datetime)):
             return NotImplemented
 
         return self._datetime > self._get_datetime(other)
 
-    def __ge__(self, other):
+    @overload
+    def __ge__(self, other: "Arrow") -> bool:
+        pass  # pragma: no cover
+
+    @overload
+    def __ge__(self, other: dt_datetime) -> bool:
+        pass  # pragma: no cover
 
-        if not isinstance(other, (Arrow, datetime)):
+    def __ge__(self, other: Any) -> bool:
+
+        if not isinstance(other, (Arrow, dt_datetime)):
             return NotImplemented
 
         return self._datetime >= self._get_datetime(other)
 
-    def __lt__(self, other):
+    @overload
+    def __lt__(self, other: "Arrow") -> bool:
+        pass  # pragma: no cover
+
+    @overload
+    def __lt__(self, other: dt_datetime) -> bool:
+        pass  # pragma: no cover
```

Comment:
Same here. Can these overloads be combined using a `Union` for the `other` type?

Reply:
same here. https://github.com/arrow-py/arrow/pull/883#discussion_r575660242